### ISSUE TITLE
[BUG] test_re_replace_repetition failed

### ISSUE THREAD

**Issue by jlowe** [NVIDIA/spark-rapids#7367]
test_re_replace_repetition failed in last night's test run:
```
def test_re_replace_repetition():
        gen = mk_str_gen('.{0,5}TEST[\ud720 A]{0,5}')
>       assert_gpu_and_cpu_are_equal_collect(
                lambda spark: unary_op_df(spark, gen).selectExpr(
                    'REGEXP_REPLACE(a, "[E]+", "PROD")',
                    'REGEXP_REPLACE(a, "[A]+", "PROD")',
                    'REGEXP_REPLACE(a, "A{0,}", "PROD")',
                    'REGEXP_REPLACE(a, "T?E?", "PROD")',
                    'REGEXP_REPLACE(a, "A*", "PROD")',
                    'REGEXP_REPLACE(a, "A{0,5}", "PROD")',
                    'REGEXP_REPLACE(a, "(A*)", "PROD")',
                    'REGEXP_REPLACE(a, "(((A*)))", "PROD")',
                    'REGEXP_REPLACE(a, "((A*)E?)", "PROD")',
                    'REGEXP_REPLACE(a, "[A-Z]?", "PROD")'
                ),
            conf=_regexp_conf)

../../src/main/python/regexp_test.py:250: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../src/main/python/asserts.py:548: in assert_gpu_and_cpu_are_equal_collect
    _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
../../src/main/python/asserts.py:479: in _assert_gpu_and_cpu_are_equal
    assert_equal(from_cpu, from_gpu)
../../src/main/python/asserts.py:106: in assert_equal
    _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
../../src/main/python/asserts.py:42: in _assert_equal
    _assert_equal(cpu[index], gpu[index], float_check, path + [index])
../../src/main/python/asserts.py:35: in _assert_equal
    _assert_equal(cpu[field], gpu[field], float_check, path + [field])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cpu = 'PROD', gpu = ''
float_check = <function get_float_check.<locals>.<lambda> at 0x7fd6cc19cb80>
path = [6, 'regexp_replace(a, A{0,}, PROD, 1)']

    def _assert_equal(cpu, gpu, float_check, path):
        t = type(cpu)
        if (t is Row):
            assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
                assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)
                for field in cpu.__fields__:
                    _assert_equal(cpu[field], gpu[field], float_check, path + [field])
            else:
                for index in range(len(cpu)):
                    _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is list):
            assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            for index in range(len(cpu)):
                _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is tuple):
            assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
            for index in range(len(cpu)):
                _assert_equal(cpu[index], gpu[index], float_check, path + [index])
        elif (t is pytypes.GeneratorType):
            index = 0
            # generator has no zip :( so we have to do this the hard way
            done = False
            while not done:
                sub_cpu = None
                sub_gpu = None
                try:
                    sub_cpu = next(cpu)
                except StopIteration:
                    done = True
    
                try:
                    sub_gpu = next(gpu)
                except StopIteration:
                    done = True
    
                if done:
                    assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)
                else:
                    _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])
    
                index = index + 1
        elif (t is dict):
            # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark
            # so sort the items to do our best with ignoring the order of dicts
            cpu_items = list(cpu.items()).sort(key=_RowCmp)
            gpu_items = list(gpu.items()).sort(key=_RowCmp)
            _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])
        elif (t is int):
            assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
        elif (t is float):
            if (math.isnan(cpu)):
                assert math.isnan(gpu), "GPU and CPU float values are different at {}".format(path)
            else:
                assert float_check(cpu, gpu), "GPU and CPU float values are different {}".format(path)
        elif isinstance(cpu, str):
>           assert cpu == gpu, "GPU and CPU string values are different at {}".format(path)
E           AssertionError: GPU and CPU string values are different at [6, 'regexp_replace(a, A{0,}, PROD, 1)']

../../src/main/python/asserts.py:84: AssertionError
```

**Comment by jlowe** [comment]
test_regexp_replace_digit also failed in a very similar way:
```
>           assert cpu == gpu, "GPU and CPU string values are different at {}".format(path)
E           AssertionError: GPU and CPU string values are different at [3, 'regexp_replace(a, [a\\d]{0,2}, x, 1)']
```

**Comment by tgravescs** [comment]
Note this is failing against spark 3.1.4 snapshot which isn't released. I'm able to reproduce this failures locally against that but not against any other spark released version so far

**Comment by tgravescs** [comment]
it appears commit https://github.com/NVIDIA/spark-rapids/commit/a42e328b7ce8c01a53722cb0f06d9b58ddc450f1 which added Databricks 11.3 is when this started to fail.

**Comment by tgravescs** [comment]
the issue is reproduceEmptyStringBug got set to true in that pr when it should be set to false.

that change originated from https://issues.apache.org/jira/browse/SPARK-39107 and it was fixed in 3.1.4 only in the 3.1.x line.
